### PR TITLE
kbnm: Fix chat messages, dev install script, whitelist published extension

### PR DESCRIPTION
**Pull request by shazow**
Before, kbnm would exec to `keybase chat send $NAME $MSG` all in the commandline, which of course would explode when $MSG was sufficiently long. This change sends the $MSG over stdin.

Also, it fixes a few bugs in the dev install script, such as a builtin `realpath` (not available on osx) and looking for `kbnm` under `$GOPATH/bin`. (Coyne/Cecile ran into these issues last week.)

Finally, it also whitelists the pubkey id of the extension in our private Chrome store listing, for testing.

### CODE REVIEW

File: go/kbnm/install_host [hunk 1]
```diff
@@ -32,13 +40,23 @@ install() {
   # Install the whitelist.
   declare target="$1"
 
-  local target_dir="$(dirname target)"
+  local target_dir="$(dirname "$target")"
```
Comment (oconnor663) [inline go/kbnm/install_host]: Nice!

File: go/kbnm/install_host [hunk 2]
```diff
@@ -8,6 +8,14 @@
 #
 # It can be run multiple times. The whitelist will be overwritten each time.
 
+realpath() {
+  if [[ $1 = /* ]]; then
+    echo "$1"
+  else
+    echo "$PWD/${1#./}"
+  fi
+}
```
Comment (oconnor663) [inline go/kbnm/install_host]: If you ever want to rewrite the script in Python, I'll support you :)

File: go/kbnm/install_host [hunk 3]
```diff
@@ -47,7 +65,9 @@ install() {
 
   chmod +r "$target"
 
-  echo "Success: Installed Chrome NativeMessaging whitelist: "$host_path" for $host_name"
+  cat "$target"
```
Comment (oconnor663) [inline go/kbnm/install_host]: Is the `cat` intentional or was it for your debugging? I'm fine either way.

File: go/kbnm/handler.go [hunk 4]
```diff
@@ -24,8 +25,21 @@ func handleChat(req *Request) error {
...
+	// super long messages.
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		io.WriteString(stdin, req.Body)
+		stdin.Close()
+	}()
```
Comment (oconnor663) [inline go/kbnm/handler.go]: Does this part need to be a separate goroutine? I think the write will return with an error, if the child happens to exit early for some reason.

File: go/kbnm/install_host (hunk 2 above)
Comment (shazow) [inline go/kbnm/install_host]: Don't think Python would buy us much over bash... Go, on the other hand... ;)

File: go/kbnm/install_host (hunk 1 above)
Comment (shazow) [inline go/kbnm/install_host]: 😊

File: go/kbnm/install_host (hunk 3 above)
Comment (shazow) [inline go/kbnm/install_host]: The old message was verbose and hard to parse, at least this way you know what got installed and easier to debug. This script is for "dev only", the consumer version will be embedded in our packaging, so I think this is better.

File: go/kbnm/handler.go (hunk 4 above)
Comment (oconnor663) [inline go/kbnm/handler.go]: And doing this on the main thread would let us return write errors more easily, not that write errors are likely. Your call either way.

Comment (shazow) [inline go/kbnm/handler.go]: Yes, the `io.WriteString(stdin, req.Body)` line blocks until it gets consumed. The stdin doesn't get consumed until `cmd.Run()` gets run below. If it isn't in a goroutine, it deadlocks.

Comment (oconnor663) [inline go/kbnm/handler.go]: Oh yeah, I guess I was imagining we replaced `Run` with `Start` -> `Write` -> `Wait`.

Comment (shazow) [inline go/kbnm/handler.go]: Ah that does make sense, good point. Let me try that.

File: go/kbnm/handler.go [hunk 5]
```diff
@@ -24,8 +25,26 @@ func handleChat(req *Request) error {
...
+	}
+
+	if _, err = io.WriteString(stdin, req.Body); err != nil {
+		return err
+	}
+	stdin.Close()
+
+	// TODO: Check/convert status code more precisely? Maybe return stdout as
+	// part of the error if there is one?
+	return cmd.Wait()
```
Comment (shazow) [inline go/kbnm/handler.go]: @oconnor663 Ping.

I think it's necessary to `stdin.Close()` explicitly (to signal the end of the stream for cmd.Wait) as well as defer'edly (alternatively litter stdin.Close() in every error branch), but it's better than a goroutine.

Thoughts?

Comment (oconnor663) [inline go/kbnm/handler.go]: Eh, I'm starting to suspect your goroutine version might save is some pain :p

I think the Most Correct Unix Thing here would be to avoid short-circuiting on those errors. So assuming `Start` succeeds, you always `WriteString`, `Close`, and `Wait` in that order, even if one of the first two return errors. Then at the end, if you did get errors, you return one of them with whatever precedence rule makes sense to you.

Comment (oconnor663) [inline go/kbnm/handler.go]: And hey, "not short-circuiting on errors" is a very well supported feature :-D

Comment (shazow) [inline go/kbnm/handler.go]: IMO checking WriteString and Close does not buy us anything, since Wait is still going to break. Pushed the version that I prefer; I think this is more equivalent to the goroutine version but without the goroutine, too.

Comment (oconnor663) [inline go/kbnm/handler.go]: That seems fine to me. If you want to put the `defer` back like you had before, that's an option too. I think a second Close fails with `EINVAL`, but we probably don't care.

Comment (shazow) [inline go/kbnm/handler.go]: I prefer the current version since it's most symmetric to the goroutine version. Gonna merge it as-is. :)
